src/Features/Core/Portable/Diagnostics/CodeAnalysisDiagnosticAnalyzerService.cs

```diff
@@ -0,0 +1,115 @@
+                _diagnosticsRefresher = diagnosticsRefresher;
+                _workspace = workspace;
+
+                _workspace.WorkspaceChanged += OnWorkspaceChanged;
+            }
+
+            private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+            {
+                switch (e.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                        _analyzedProjectIds.Clear();
+                        break;
+                }
+            }
+
+            public bool HasProjectBeenAnalyzed(ProjectId projectId) => _analyzedProjectIds.Contains(projectId);
+
+            public async Task RunAnalysisAsync(Solution solution, Action<Project> onProjectAnalyzed, ProjectId? projectId, CancellationToken cancellationToken)
+            {
+                if (projectId != null)
+                {
+                    var project = solution.GetProject(projectId);
+                    if (project != null)
+                    {
+                        await AnalyzeProjectCoreAsync(project, onProjectAnalyzed, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    var tasks = new Task[solution.ProjectIds.Count];
+                    var index = 0;
+                    foreach (var project in solution.Projects)
+                    {
+                        tasks[index++] = Task.Run(() => AnalyzeProjectCoreAsync(project, onProjectAnalyzed, cancellationToken), cancellationToken);
+                    }
+
+                    await Task.WhenAll(tasks).ConfigureAwait(false);
+                }
+            }
+
+            private async Task AnalyzeProjectCoreAsync(Project project, Action<Project> onProjectAnalyzed, CancellationToken cancellationToken)
+            {
+                await _diagnosticAnalyzerService.ForceAnalyzeProjectAsync(project, cancellationToken).ConfigureAwait(false);
+                _analyzedProjectIds.Add(project.Id);
+                onProjectAnalyzed(project);
+                _diagnosticsRefresher.RequestWorkspaceRefresh();
+            }
+
+            // Running code analysis on the project force computes and caches the diagnostics on the DiagnosticAnalyzerService.
+            // We return these cached document diagnostics here, including both local and non-local document diagnostics.
+            public Task<ImmutableArray<DiagnosticData>> GetDocumentDiagnosticsAsync(DocumentId documentId, Workspace workspace, CancellationToken cancellationToken)
+                => _diagnosticAnalyzerService.GetCachedDiagnosticsAsync(workspace, documentId.ProjectId,
+                    documentId, includeSuppressedDiagnostics: false, includeLocalDocumentDiagnostics: true,
```

Comment:
doc what the false/true args are for.  in particular, i don't know what a local vs non-local diag is.

Reply:
I have added `param` doc comments to all the relevant methods on `IDiagnosticAnalyzerService` to clarify what these flags indicate.